### DISCUSSION TITLE
Lazy "new"

### DISCUSSION THREAD

**Discussion by spoerri** [jj-vcs/jj#4012] [Ideas]
It would be nice to banish empty commits, and to enable users not to have to care about them in the form of \@- vs \@. It would also be nice to be a little more opinionated and automatic for "squash workflow" vs "edit workflow".  For example, users shouldn't have to care about whether to use `new` or `edit` for modifying vs viewing historical commits. #3129 #3660

Therefore, let's make `new` be lazy in the sense that a commit/change is only created when the working copy is actually modified.

I.e. for: 

    1. jj new [someExistingRev or not]
    2. ...add, modify or delete a file in the working copy...
    3. ...any jj command...

This proposal is to create the commit at 3 instead of 1.

Other niggles: would probably also want to go ahead and create the commit for `jj describe` modifications as well as working copy modifications. Might also want to rename `new` to something like "get", especially if there's openness to employ `commit` instead of `new` in one use case.

(And fun: to one way of looking at it, this would make the old "open" vs "closed" be a property of the working copy! The working copy would open either via `edit` or on a commit which it just created, and would close with `new`, `commit`, or `edit` of a different commit.)

**Comment by ilyagr** [comment]
> let's make new be lazy in the sense that a commit/change is only created when the working copy is actually modified

I do sometimes want to `jj edit` a commit, especially if it's a WIP commit with no children. I'm not sure how to fit this in with a "lazy" new.

Also, this was just before my time, but `jj` did formerly have an open/closed concept, and then it was simplified away. See https://github.com/martinvonz/jj/discussions/321#discussioncomment-2788147. I never learned precisely how that worked, but I imagine it was at least somewhat similar to what you are thinking.

**Reply by spoerri** [reply]
> I do sometimes want to `jj edit` a commit

Yes, that shouldn't change. On the repo level, only change is to not have empty commits. On the UI level, only change is that people shouldn't ever feel like they have to use `jj edit` to avoid creating empty commits.

> open/closed concept [...] at least somewhat similar to what you are thinking

I remember! The original open/closed was a _commit_ concept; I thought it was cute that this is a _working copy_ concept. And users shouldn't actually have to think about open and closed. Perhaps it's just a distraction from the proposal ☹️

**Reply by ilyagr** [reply]
It seems pretty important to the proposal to be able to distinguish the state after `jj edit aaa` and after the lazy `jj new aaa` somehow. So, I think a notion of closed/open working copy, or something analogous, would have to be a part of it.

I think the status quo is pretty elegant in how it uses an empty commit to avoid needing such a notion, so that's an advantage any more complex design would have to overcome. However, I don't want to rule out the possibility of a better design; the empty commits are a bit of a nuisance (though Martin recently made them less likely to persist after you navigate away, even for merge commits); perhaps there is a way to add a notion in a way that makes the UI simpler.

**Reply by spoerri** [reply]
> a notion of closed/open working copy, or something analogous, would have to be a part of it

I hear. There would be an extra boolean, let's call it isEditing. It would _only_ be used for determining whether the lazy-new should be triggered. When it is false and a modification is detected, jj would create a new change and associate it with the working copy and set isEditing to true. The `jj edit` command would set it to true immediately. It should be exposed in the `jj status` output, but I don't think people would care much about it. `jj commit`, `jj squash`, etc would reset isEditing to false instead of creating an empty commit.

The advantage is that **every command operates on the one obvious commit that you're looking at** by default. You don't have to worry about whether you called commit, or whether any particular command expects to be called before or after commit.

**Reply by ilyagr** [reply]
>  It should be exposed in the jj status output, but I don't think people would care much about it. 

I would want to know whether editing a file will create a new commit. I'd want this piece of the working copy state to be shown in my prompt and in `jj log` next to the working copy. I don't think trying to hide or deemphasize this boolean would help.

**Reply by joyously** [reply]
Where is the advantage of this? How does this work when you interleave `jj` and `git` commands?

**Reply by spoerri** [reply]
> Where is the advantage of this?

It’s more WYSIWYG and in the spirit of jujutsu: the working copy is king. All simple commands would operate on what’s in the working copy now, with no additional step or state (e.g. commit command or empty commit) needed to get them to do the right thing.

Lazy-new achieves this because it’s a good balance between jj policy and user intent. When the user names a revision, he wants to work with that revision until there’s a good reason to work with something else. Jj (and perhaps commonsense) dictates that historical versions should generally not be modified directly. And the user wants to be able to modify the working copy and have the right thing happen. Lazy-new satisfies all those constraints. If a new change is created just when it is needed instead of manually in advance, the current change is always the right one, and tracking whether the user is “done with” a change is unnecessary complexity which can be jettisoned. Perhaps better branding would be to call the design “just in time” new instead of “lazy” new. Ah well.

(Obviously a transition to that full design would be more disruptive than just adding a lazy-new feature and simplifying nothing else. That end design is the real motivation though. Could talk later about how to reasonably get there, e.g. experimental config…)

> How does this work when you interleave jj and git commands?

I’d imagine that interleaving jj and git commands would be more or less the same as now: git doesn’t know about anything new in the working copy since the last jj command. So if “new” is no longer a jj command run in advance, when there is a fresh change sitting in the working copy, git wouldn’t see any sign of it. In the current approach, git would only see an empty commit, which doesn’t seem any more useful. When a jj command was most recent, git would see exactly the same in both approaches. But I don’t have experience with this, so let me know what I’m missing!

@ilyagr mentioned above empty commits being made:
> less likely to persist after you navigate away

So perhaps just-in-time-new is just a more dramatic and simple step along that path? Instead of automatically dropping empty commits, we could not create them in the first place.

**Reply by joyously** [reply]
> All simple commands would operate on what’s in the working copy now, with no additional step or state (e.g. commit command or empty commit) needed to get them to do the right thing.

Could you list the commands which right now have additional step or state to do the right thing?
> And the user wants to be able to modify the working copy and have the right thing happen.

Can you list where this is not the case now?
>  git doesn’t know about anything new in the working copy since the last jj command. 

Neither does `jj`, unless there is a daemon.
>  In the current approach, git would only see an empty commit, which doesn’t seem any more useful.

Would you say that the empty commit doesn't hurt anything either? To me, it's useful to have a consistent result from a command. Every time I say `jj new`, I should get a new change, whether empty or not.
Your argument sounds more like you don't want the perspective to change from "store my finished work" to "save as I work".
If I may quote the README, here is why the command needs to be consistent:
- Changes to files are recorded automatically as normal commits, and amended on every subsequent change. This "snapshot" design **simplifies the user-facing data model** (commits are the only visible object), **simplifies internal algorithms**, and completely subsumes features like Git's stashes or the index/staging-area.
- Operation log & **undo**
- Safe, **concurrent replication**

**Reply by spoerri** [reply]
Forgive me for not responding to all points. At work, at some point along the way, we'd jump into a conference room instead of going back and forth. I'll just respond to your request to "list the commands which right now have additional step or state to do the right thing"...

When I want @:

- when I have a change "in progress" checked out

When I want @-:

- when i'm working with a historical revision (unless i checked out with `jj edit` which is not recommended)
- after I call `jj commit` but before i make another change
- `jj branch create` (and the same with tags when they become available)
- `jj git push`

I'm suggesting that a change being "in progress" shouldn't matter to any command. If I see a change in front of me, I want commands to apply to that (\@), and a subsequent empty commit to signal "the change that you see is no longer in progress" only interferes with that.

I probably won't come back to this thread soon. I'm really just curious what @martinvonz thinks...

**Reply by joyously** [reply]
Yes, we all want to know what Martin thinks. (I hear he's on vacation.)
Thank you for stating the problem, instead of starting with a "solution" and working backward.
I think you listed a couple of things (Git-related) that want @- but most things use @.
My bias may show through, but let me try to state what I see a VCS doing.
1. Store my work in a way that's easy for others to build on.
2. Show me the difference between old version and new version.
3. Show me history of changes.
4. Extract historical version (so I can test other's work or start over).

Above all, the tool should keep things simple, so the human doesn't have to keep track of things.
I think it would reduce mental overhead if historical version extraction happened external to the working copy.

**Reply by martinvonz** [reply]
This has come up several times before. I haven't seen a proposal that seem simpler than what we have yet. If we don't eagerly create an empty commit, then when do we create it? Lots of commands currently default to using `@` as their argument. If `@` doesn't exist until some later point, then we'll have to figure out what to use instead. For example, `jj diff` defaults to `@`, so if you `jj new foo; jj diff`, then that's going to show an empty diff. What will it do if we lazily create `@`? Will it still show an empty diff because we create `@` right when `jj diff` is called? Then it seems we're going to create the commit in almost every case anyway, so it makes no practical difference from our current solution of eagerly creating the commit.

**Reply by amiryal** [reply]
I think I understand where @spoerri’s frustration is coming from. Assuming adherence to the squash workflow, when one wishes to check out a certain revision, there is a need to think about whether to use `jj new` or `jj edit`. This is made more intimidating by the fact that `jj new` creates extra commits, which may feel unnecessary and wasteful if you don’t treat JJ as a “black box”.

My suggestion, in order to avoid the dilemma between `jj new` and `jj edit`, would be to **revive the `jj checkout` command with a twist**. Instead of always creating a new empty change like `jj new`, it would:
* Behave like `jj edit` if the target is a leaf commit which is either empty or has no description.
* Behave like `jj new` if the target commit either has descendants or is non-empty with a description.

For checking out branches pointing to the latest “committed” changes, this may be further enhanced with:
* Behave like `jj edit TARGET+` if the target commit has exactly one descendant which is either empty or has no description.

If you come from Git and know how `git stash` works, then it may help to think of JJ working copy commits as automatic Git stash commits.

**Reply by arxanas** [reply]
> The advantage is that every command operates on the one obvious commit that you're looking at by default. You don't have to worry about whether you called commit, or whether any particular command expects to be called before or after commit.

This very advantage seems to be negated by the proposal at hand? If you do lazy-new, then you do _not_ know ahead of time whether the commit you most recently checked out is in fact still the commit that's checked out, because it depends on operations that happened outside the version control system since then.

In my opinion, one of the major advantages of the jj workflow in comparison to Git is that you don't have to constantly think about repo context when conducting operations, such as the contents of the staging area or whether a rebase/merge is in progress. In this case, lazy-new would add extra context: whether a new commit has actually been created since I last looked at the commit graph.

I would propose this invariant: if you haven't explicitly run a jj command that modifies the structure of the commit graph, then it should remain isomorphic to your last-known state of the commit graph. This is satisfied by `new` but not lazy-new.

**Reply by spoerri** [reply]
> If we don't eagerly create an empty commit, then when do we create it?

Yes, we potentially create a commit on every jj command only _when the working copy has been newly  modified_.

The terminology hints that this is right: what do we call an evolvable commit? A **change**! It’s fitting that jj should give you change object exactly when you make a change! 😄 

> If @ doesn't exist until some later point

\@ should always exist. If the user has not made any changes, \@ is the originally specified commit. So `jj get someRev; jj diff` would show the diff between someRev- and someRev and `jj get someRev; echo foo>>bar; jj diff` would show the change to bar.

**Reply by spoerri** [reply]
> If you do lazy-new, then you do not know ahead of time whether the commit you most recently checked out is in fact still the commit that's checked out

What's the practical ramification of this? When you're viewing historical commits, you might not realize that you've made changes in the working copy, but that's true of either design. And unless you just abandon those commits, you're going to want to figure it out with `jj st`. And that's also true in both designs.

**Reply by arxanas** [reply]
> What's the practical ramification of this?

I always use `jj new`. If I'm investigating a historical commit, I might want to make changes/do builds/etc., but not be sure if I've left changes in the working copy. I might want to run e.g. `jj diff --from X --to @-` to determine something about the commit I'm investigating. If I've made working copy changes, then I might get silently very-slightly-wrong results with lazy-new. Of course I can run `jj status` before all commands like that, but the point is to reduce the amount of context I have to check before various operations.

**Reply by spoerri** [reply]
OK, I hear. Yes, lazy-new would remove the ability to easily and reliably exclude changes that might have crept into the working copy, which is useful for when you're just going to abandon the working copy anyway. Is it just for `jj diff`?

If that's sufficiently useful, there are ways to support it without such a global impact on the design.

> I always use jj new

Yes, the whole motivation for this proposal is that "new" is the recommended mode of operation. If we were happy with edit workflow, we could go do fine with neither "eager" nor "lazy" new; we could go with "explicit only" new.

It does raise another point in the design space: instead of creating an empty commit, we could make the working copy readonly. The only supported ways to make it writable would be `jj new` and `jj edit`. Forked a different discussion for that: #4047

**Reply by arxanas** [reply]
> Is it just for `jj diff`?

In practice, I believe I use it for a variety of commands, like `jj rebase`, `jj split`, `jj restore`, etc., without wanting to check `jj status` first.

That is, I think there are a lot of workflows that I personally use that depend on `@` being "stable" in a certain sense.

> If we were happy with edit workflow, we could go do fine with neither "eager" nor "lazy" new; we could go with "explicit only" new.

As far as I can tell, there are in fact two factions of users, where one strongly prefers `jj new` by default, and one strongly prefers `jj edit` by default, and neither subsumes the other.

- In that you couldn't forgo one and express all of your typical operations with it using the other one.
- I don't think lazy-new subsumes either one, either?

> It does raise another point in the design space: instead of creating an empty commit, we could make the working copy readonly. The only supported ways to make it writable would be jj new and jj edit.

In a sense, this is kind of like preserving a staging area idea, and not really committing the working copy changes to the commit graph until an explicit "commit" operation of some sort.

While I think this idea would still be executed better in jj than in Git, since we would still treat the working copy as a normal kind of commit and use all the same tools to operate on it, we really want to reduce the number of logical states and transitions that the user has to think about.

- [*What's wrong with Git? A conceptual design analysis* (2013)](https://spderosso.github.io/onward13.pdf) gives some of the examples and arguments towards reducing the number of states/transitions.
  - It doesn't include extra other state like ongoing rebases/merges in the figures.

> If that's sufficiently useful, there are ways to support it without such a global impact on the design.

My feeling so far is that the jj-new model is indeed a local maximum in the design space, but I don't think I draw the same conclusions about how to evolve the design as you?
  - It's not necessarily ideal, and one has to think about `@` vs `@-` maybe more often than one would like, but I think limiting the number of design concepts here has been worth the reduction in complexity.
  - My intuition is that an approach to change the UI around the existing jj concepts would be more successful than literally changing the jj semantics.
    - For example, the suggestion of adding a specific symbol to indicate `@` has come up several times. It adds no new expressivity to jj itself, but maybe it turns out to be a significant UI improvement in practice?
- But I don't see that lazy-new or default-read-only addresses the design problems in the right way.

---

If we change the jj design, since I think the current design occupies a local maximum, I suspect that we would need to do a less incremental step in the design space to arrive at a better solution.
- My own proposal was [hierarchical commits](https://github.com/martinvonz/jj/discussions/321#discussioncomment-2832646),
  - which generalizes the "open/closed" concept from $n \in {0, 1}$ to $n \in \mathbb{Z^+}$.
- While we may never implement hierarchical commits, the interesting thing would be tackling the problem of *grouping commits arbitrarily*.
  - Under hierarchical commits, we would always create new working copy commits, even when "editing" historical commits, and the problem becomes grouping them meaningfully instead.
  - But the same grouping problems exist today. For example, branches, [topics](https://github.com/martinvonz/jj/issues/3402), and [stacks](https://github.com/martinvonz/jj/issues/3713) are all solutions that aim at the problem of grouping commits meaningfully.

What we can take away is the UI problem of whether the resulting working-copy commit is *grouped with the parent commit or not*.
- This problem manifests itself in both
  - how we physically render it in the commit graph — quite important when it comes to onboarding users and building their intuition,
  - and also how we address the commit versus its hierarchical ancestors programmatically.
    - Currently `@` is the only such option.
    - There is not currently a way to delineate/separate the working-copy commit from its parent commit in a semantically-meaningful way in the revset language, or, really, the jj design itself. It's all just convention, for which we generally want to use `@-` to address the parent commit, but not always.
- [Tangential] To express grouping in the hierarchical commits model,
  - we would require an operation in terms of some $n$ expressing hierarchy:
    - it would be some kind of term meaning "the chunk of this commit graph that spans from the previous 'fork point' to the next 'merge point'",
    - and the ability to express " $n$ = how many levels of hierarchy to traverse upward when finding the desired fork- and merge-points".
  - `jj new X` would be like adding a new commit after `X` inside its parent commit, and addressing that new commit with $n = 0$.
  - `jj edit X` would be like adding a new commit *inside* `X`, but continuing to address `X` by using $n = 1$.
    - Realistically, I think users of an "edit workflow" might prefer to use $n = 0$ in many cases.
  - `jj lazy-new X` would be something like adding the new commit inside `X` (like `jj edit`), but operations use $n = 0$ in the specific case when jj wants to "instantiate" the lazy commit with the working-copy changes, or $n = 1$ otherwise.
    - Critically, the user could still choose to explicitly address the working-copy commit with $n = 0$ or parent commit with $n = 1$ if they so desired, which would support my workflows mentioned previously.
    - The design complexity is now *addressing* commits, not creating them. The commit graph is now "stable" in the sense I alluded to.
  - Branching becomes a matter of groups with $n > 1$.

To summarize:
- If we're going to change something, then we may prefer to change the entire way that jj organizes/groups commits, rather than handle just the case of the working-copy commit differently.
- The entire workflow may well be best addressed a UI problem, not a fundamental design problem.

**Reply by arxanas** [reply]
Another idea is to have some kind of symbol that means "the working-copy commit if non-empty, or the parent commit if empty". Then people who use the "lazy-new" workflow can enjoy that today. Actually, I bet we could express that right now with a revset alias? (We just wouldn't have a cool symbol for it.) Then the main problem is fixing most commands to operate on *that* (or some other user-definable revset that means "the thing I'm working on"). @spoerri I'm curious if that addresses your workflow issues?

**Reply by spoerri** [reply]
> symbol that means "the working-copy commit if non-empty, or the parent commit if empty" [...] addresses your workflow issues?

maybe... i was really hoping to _reduce_ complexity, so adding features without replacing anything is a risky way to achieve that.

you probably saw in the other discussion that i'm settling on `jj commit` as the problem.

i've been conflating "central" and "leaf" empty commits. i'm coming around to the idea that leaf empty commits don't really have much conceptual impact. OK, it's true that users can't simply do `jj checkout somHistoricalRev; jj branch create X` but `jj checkout somHistoricalRev; jj branch create X -r somHistoricalRev` isn't so bad if you don't want to think about \@ and @-; you have to know which revision you checked out anyway. And once you switch to another commit, those leaves are fleeting, so to say.

i'm more concerned about standardizing on "finished" commits being only those with child commits. assuming the advance branches stuff goes the right away, maybe all i want is a calling convention for `jj git push` that requires neither running `jj commit` nor specifying weird parameters.

**Reply by lgarron** [reply]
> Another idea is to have some kind of symbol that means "the working-copy commit if non-empty, or the parent commit if empty".

I found this issue on GitHub search while trying to figure out how to make my own local flows more reliable. I'm currently very tempted to write myself a script to compute a way to refer to exactly this concept.

It is very common that I want to check out a commit, maybe perform some work/run some tests on it (or maybe leave it unchanged), and update + push the `main` bookmark from "the latest commit". Sometimes `jj edit` was the most appropriate approach to check out that commit, sometimes `jj new` was necessary.

This means that I have to sometimes move the bookmark to `@` and sometimes to `@-` depending on whether the commit happens to be immutable — which can happen due to entirely unrelated reasons (e.g. whether the given commit happens to be part of another pushed branch).

I don't mind looking at the commit graph to orient myself, and I appreciate that [`gg`](https://github.com/gulbanana/gg) lets me move the `main` bookmark very intuitively. But I would really like to be able to use a flow that does not require me to "look over my shoulder" every single time I want to push a branch.

**Reply by steveklabnik** [reply]
@lgarron you might be interested in various versions of an alias called `tug` that people use, here's one thread about it https://github.com/jj-vcs/jj/discussions/5568

**Reply by lgarron** [reply]
@steveklabnik: ooh, thanks, very helpful! 🤩

**Reply by jennings** [reply]
> This means that I have to sometimes move the bookmark to `@` and sometimes to `@-` depending on whether the commit happens to be immutable

If you can really boil it down to that rule, you could make a revset alias to help, something like this:

```toml
'here' = 'coalesce(@ ~ empty() ~ immutable(), @-)'
```

Which evaluates to `@` if it's mutable and non-empty, and `@-` otherwise.

**Reply by lgarron** [reply]
@jennings: Thanks, that was a really helpful starting point, and I'm definitely gonna try it.

A small detail is that I think going back to `@-` when `@` is immutable (and non-empty) doesn't make sense to me[^1]. But after reading up on the revset language, I think this works?

```
'here' = 'coalesce(@ ~ empty(), @-)'
```

[^1]: If I am editing a commit using `--ignore-immutable` then, well… I'm ignoring immutability already.

**Comment by joyously** [comment]
> This proposal is to create the commit at 3 instead of 1.

Despite you suggesting to rename `new`, I think the command should actually do its job when invoked. No one can predict the order or timing of commands to come. The next command might be from a different user, months after, or a tool, milliseconds after. Making a new commit is the expected result, or the command would not be used. It should be consistent, or it muddies up the mental model.
Is there some penalty for an empty commit?

**Reply by spoerri** [reply]
IMHO not aggressively creating commits is the correct mental model! I didn’t lead with that, lest I cause a tumult.

From a practical perspective, empty commits are not the end of the world. So far I’ve just seen trivial issues around \@- and the correct semantics for `jj next`. To be clear, I don’t mean literally `jj new` but any command that creates a new commit in advance. `jj commit` is the most egregious. This is the sort of thing that increasingly many things come to rely on.

It’s more important to avoid a design that relies on empty commits because they are a smell in the model. We employ empty commits now as a replacement for “the parent commit is closed”, but the real problem with open/closed is that it doesn’t belong in the commits at all. It belongs in the working copy.

Multiple working copies ala workspaces might demonstrate: you can edit a commit in one workspace and view it in another. So having an empty commit as its child doesn’t accurately represent reality.

Or maybe we don’t even need to look at such an esoteric case; it’s just wrong to have users create new commits in order to view historical revisions. (Of course it’s also unfortunate to use `jj edit` for that.) IMHO jujutsu has great potential to “scale down” and be simple for newbies. So it’d be lovely to be able tell users “you can open up any revision with: `jj get`, and if you make any changes jujutsu will automatically and safely save them.” It’s hard to promote jujutsu as a simplified git if some of the most basic use cases require explaining revsets and squash vs edit workflows. Meet the new revision control same as the old revision control. Anyway, sorry to rant 😄

**Reply by joyously** [reply]
> avoid a design that relies on empty commits because they are a smell in the model. We employ empty commits now as a replacement for “the parent commit is closed”, but the real problem with open/closed is that it doesn’t belong in the commits at all. It belongs in the working copy.

I don't think you have described the entire model. In my understanding, no commit is "closed", while the working copy as commit simplifies the interface and reduces possibility of lost work.

> you can edit a commit in one workspace and view it in another. 

If it's two workspaces, it's not the same commit? I can easily see how you can be editing a commit and view it in another window (no extra workspace needed). That's always the nature of multitasking. I just don't see what the contents have to do with it.

> it’s just wrong to have users create new commits in order to view historical revisions. 

I agree. I don't think they have to, however. There are other commands besides `new`.

> So it’d be lovely to be able tell users “you can open up any revision with: jj get, and if you make any changes jujutsu will automatically and safely save them.”

I think this would be a big problem. I wrote up a similar sounding [issue](https://github.com/gulbanana/gg/discussions/5#discussioncomment-9021865) while I was testing out `gg` and it seemed very fragile if a novice could be browsing the repo, leaving a trail of changes from editor autosaves.

> It’s hard to promote jujutsu as a simplified git if some of the most basic use cases require explaining revsets and squash vs edit workflows. 

It's not a simplified git. The most basic use case doesn't need revsets or squash or edit either. Most basic is change files, then `jj new`. You don't even have to describe it. You can always simply march forward and ignore most of the commands.

**Comment by scott2000** [comment]
I personally prefer the current model because it feels very consistent since `@` is always present and always refers to the working copy (meaning that it's empty if there are no changes made yet). I also wouldn't like the commit graph to change due to filesystem changes.

Maybe some of these issues could be resolved more easily by just adding a new revset called `#` which just means `@-`? So the "checked-out" commit would just be `#` always, and the working-copy commit would remain as `@`, and you could just configure `jj log` to hide `@` if there's no changes so that you'd only see `#`. This has the benefit that all of the commands would still work the same as they do now (e.g. `jj describe` could still mean `jj describe @`).

So basically when you check out a commit, it would show that commit as `#` in the log (and the empty `@` would be hidden). Then if you make changes, now `@` would show up on top of `#`. But behind the scenes, the working copy would still always be accessible using `@` regardless of if it's visible in the log.
